src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/RecursiveSpinLockTest.kt

```diff
@@ -0,0 +1,502 @@
+    ) = array.cas2(index1, expected1, update1, index2, expected2, update2, 0)
+
+    @Operation
+    fun cas2_1(
+        index1: Int, expected1: Int, update1: Int,
+        index2: Int, expected2: Int, update2: Int,
+    ) = array.cas2(index1, expected1, update1, index2, expected2, update2, 1)
+
+
+    @Test
+    fun modelCheckingTest() =
+        ModelCheckingOptions()
+            .addCustomScenario {
+                parallel {
+                    thread { actor(BrokenCas2RecursiveLiveLockTest::cas2_0, 0, 0, 2, 1, 0, 3) }
+                    thread { actor(BrokenCas2RecursiveLiveLockTest::cas2_1, 0, 0, 4, 1, 0, 5) }
+                }
+            }
+            .iterations(500)
+            .invocationsPerIteration(1000)
+            .sequentialSpecification(IntAtomicArraySequential::class.java)
+            .checkImpl(this::class.java)
+            .checkLincheckOutput("spin_lock/broken-cas-2-recursive-live-lock.txt")
+}
+
+/**
+ * Broken CAS-2-based array implementation with only 2 operations. Calling them may lead to alternation recursive live-lock.
+ */
+class AtomicArrayWithCAS2<E : Any>(size: Int, initialValue: E) {
+    private val array: AtomicArray<Descriptor?> = atomicArrayOfNulls<Descriptor>(size)
+    private val gate0 = atomic(false)
+    private val gate1 = atomic(false)
+
+    init {
+        // Fill array with the initial value.
+        for (i in 0 until size) {
+            array[i].value = Descriptor(i, initialValue, initialValue, i, initialValue, initialValue).apply {
+                status.value = Status.SUCCESS
+            }
+        }
+    }
+
+    fun cas2(
+        index1: Int, expected1: E, update1: E,
+        index2: Int, expected2: E, update2: E,
+        turn: Int
+    ): Boolean {
+        require(index1 != index2) { "The indices should be different" }
+
+        val descriptor = if (index1 < index2) Descriptor(index1, expected1, update1, index2, expected2, update2)
+        else Descriptor(index2, expected2, update2, index1, expected1, update1)
+
+        descriptor.apply(turn = turn)
+        return descriptor.status.value == Status.SUCCESS
+    }
+
+    private inner class Descriptor(
+        private val index1: Int,
+        private val expected1: Any?,
+        private val update1: Any?,
+        private val index2: Int,
+        private val expected2: Any,
+        private val update2: Any
+    ) {
+        val status = atomic(Status.UNDECIDED)
+
+        fun read(index: Int): Any {
+            check(index == index1 || index == index2)
+            return if (status.value == Status.SUCCESS) {
+                if (index == index1) update1 else update2
+            } else {
+                if (index == index1) expected1 else expected2
+            }!!
+        }
+
+        /**
+         * @param broken if true then we never will go out of recursive spin-cycle
+         */
+        fun apply(initial: Boolean = true, turn: Int = -1, broken: Boolean = false) {
```

Comment:
Why do you need `broken: Boolean = false` flag here?
Maybe it is better to just split this method into two methods, with one of them being "broken" (and its name should reflect this, like `applyBroken`).

Reply:
In general, could you please document this test, since it is not trivial. 
In particular, to explain why you need two versions of `apply` and `installOrHelp` methods (broken one and the correct one).